Restore the Cygwin port of OCaml

dra27 opened the pull request:
The first three commits restore the Cygwin port and ready for feedback. In particular, Cygwin's `mmap` cannot be used in the same way as other Unix systems both because it's ultimately backed by `VirtualAlloc` (so can't be used to trim existing regions) and also because there's an unimplemented feature. The reservation part of the minor arena therefore uses `mprotect` instead. These functions in `platform.c` are becoming just a little `#ifdef`-y, though.

With those first three commits, the entire testsuite passes _except_ for `tests/regression/pr9326/gc_set.ml` - i.e. the one test which resizes the minor arena. The fourth commit temporarily works around this problem by having it that on minor arena resize, we decommit the minor heaps but don't `munmap` the memory itself. I've tried debugging this a bit (see [this cygwin list message](https://cygwin.com/pipermail/cygwin/2022-October/252326.html)).

The remaining commits temporarily allow for testing in this PR and are not intended to be merged. While debugging this, I added GC log level for 0x1000 for the memory mapping subsystem - if this sounds useful, I'd be happy to complete that with the Windows side and put it in a separate PR.

dra27 committed:
Re-enable Cygwin

dra27 committed:
Missing CAMLextern in caml/runtime_events.h

dra27 committed:
Cygwin caml_mem_* implementation

Cygwin's mmap is ultimately backed by the Windows implementation, so we
can't rely on being able to trim portions of memory. Additionally,
Cygwin's mmap doesn't support changing the protection of committed pages
- it requires mprotect to be used for this.

dra27 committed:
AppVeyor script fixes for Cygwin

dra27 committed:
Show timings on precheck

dra27 committed:
Add GC log level 0x1000 for memory reservations

dra27 committed:
Require memory chunks to be whole pages

Instead of rounding the size, caml_mem_map now requires the caller to
ensure that size is a multiple of caml_sys_pagesize. The alignment is
still permitted to be less.

A similar assertion is added to caml_mem_commit.

gasche commented on runtime/platform.c:
```diff
@@ -154,10 +157,15 @@ static struct lf_skiplist mmap_blocks = {NULL};
   void* mem;
   uintnat base, aligned_start, aligned_end;
 
+  CAML_ASSERT_SIZE_IS_PAGE_ALIGNED(size);
```
Why not `CAMLassert(Is_page_aligned(size))`, following the Is_power_of_2 style above?

dra27 replied:
Much better that way, yes 🤦

gasche reviewed (commented):
Indeed I fear that `caml_mem_map` is becoming too hard to read. (It was already hard.) Could we try to separate the system-specific code in separate functions?

I see two general approaches:
- have three implementation of the whole function (Unix, Cygwin, Windows), factoring non-system-specific logic in sub-functions
- have a common implementation that calls to system-specific routines

(of course a mix of the two strategies is possible.)

I think we could have something as follows:

```c
void* caml_mem_map(size, alignment, reserve_only) {
  <alignment stuff, alloc_sz etc.>;
  
  do { /* some architectures need a retry loop */
    mem = os_reserve_memory(..);
    aligned_start = ...;
    aligned_end = ...;
    mem = os_trim_reserved_area(mem, aligned_start, aligned_end); /* NULL on failure */
  } while (mem == NULL);
  
  <insert in debug skiplist>;
}
```

gadmm commented:
I see that you explained in a comment the point about being unable to call VirtualAlloc directly.

I think it looks better to have separate implementations. One main function (with the common logging and debugging code) and one platform-specific sub-function should be enough.

dra27 committed:
Factor out platform-specific memory management

mmap-based versions -> unix.c; VirtualAlloc-based version -> win32.c

dra27 commented:
I rebased to eliminate the merge conflict, but I've pushed the changes themselves as additional commits for now. Thanks for the feedback (also in #11645) @gasche and @gadmm. I had missed the opportunity to take advantage of `dwAllocationGranularity` for the Windows port (nice bit of history in [The Old New Thing](https://devblogs.microsoft.com/oldnewthing/20031008-00/?p=42223) on why that's portable and likely to stay so, even with the departure of those older architectures).

I changed `caml_sys_pagesize` to be `intnat` rather than `long` in preference to having to add `ARCH_LONG_PRINTF_FORMAT` definitions.

I wasn't convinced it was going to work with the earlier versions, but I think that factoring out the Windows `VirtualAlloc`-based and Unix `mmap`/`mprotect` based versions to their respective platform-specific objects makes `platform.c` look a bit neater.

gasche commented on runtime/win32.c:
```diff
@@ -1113,3 +1113,32 @@ int64_t caml_time_counter(void)
+  return
+    VirtualAlloc(NULL, size,
+                 MEM_RESERVE | (reserve_only ? 0 : MEM_COMMIT),
+                 reserve_only ? PAGE_NOACCESS : PAGE_READWRITE);
```
I don't think the phrasing of the comment is right (or I misunderstood something): the comment suggests that `alignment` is always `caml_sys_pagesize`, which is not correct I believe when allocating major-heap pools. The remark of @gadmm was that the (larger) alignment of major-heap pools (32KiB) remains smaller than the (pretty large) alignment of VirtualAlloc (64Kib).

This property that `alignment` is always smaller that VirtualAlloc's actual alignment is not guaranteed by the runtime configuration (eg. choice caml_sys_pagesize), it is a dynamic property that can only be verified by checking all callers of `mem_map`. So I think that there should be a dynamic check in the function here to ensure that it holds (even outside the debug mode).

dra27 replied:
It may very well be both 🙂 The comment has lost a bit of context in being refactored, because the fact that `alignment` has already been rounded up to `caml_sys_pagesize` is not apparent (I'll clarify that). _But_ the fact that `caml_mem_map` should not be used on Windows or Cygwin with `alignment > caml_sys_pagesize` indeed is not mentioned anywhere.

A key part of the simplification here (getting rid of the retry loop, etc.) is that we're _not_ providing a general reservation function anymore - the Windows implementation is capitalising on the knowledge that we use `caml_mem_map` in two places, one of which is always "page"-aligned (minor heaps) and the other of which we can statically check is page aligned (just about, for the major pools - the 64KiB page size on Windows is basically a fact, and `POOL_WSIZE` is fixed at compilation time). Is a comment at the two call sites (in `shared_heap.c` and `domain.c`) more appropriate than an assertion? Or a header assertion (as used for domain_state) that `POOL_WSIZE` is is `<= 64KiB`?

gasche replied:
I think that a dynamic check would be good to have, because:
- it has no performance impact: the check is basically free, and this function is called infrequently anyway
- it is the only way to guaratee that this function makes sense in isolation, without inspecting its callsites
- it will help us detect that we are doing something wrong if we add callsites later that break this assumption, or change existing callsites in a breaking way (for example moving to 128Kib pools).

In other words: I agree that we are not providing a general function anymore, but you can still give a much more general and usable spec that "usable in only two places in the runtime", which is "usable for any alignment value that divides 64Kib", and defensively checking this precondition is good programming practice.

We could also comment at the callsites if we wanted (comment to justify why the dynamic check will not fail for this call), and we could have a static assertion that POOL_WSIZE divides 64KiB. But I would still prefer to have a dynamic check in addition.

gasche replied:
Re. `caml_sys_pagesize`: my first comment above was written in a weird mental model where `sys_pagesize` is 4Kib on all systems, and here we are talking about a distinct, larger alignment value. If you configured `pagesize` to be 64kiB on Windows, then what I wrote a first does not really make sense. I think this is what you did in the current iteration with the PR, cheating a bit by conflating the notion of "page size" (for committed memory) and "granularity" (for virtual address space reservation). I find the resulting situation, uh, a bit confusing (the internet agrees that "page size" on Windows is 4Kib or 8KiB, and here we use something fairly different).

gadmm replied:
Reservation granularity is 64kiB but committing granularity is 4kiB, so there should be no need to change any runtime size. I agree this should just be a matter of throwing an error if an alignment >64kiB is requested.

gasche replied:
No runtime sizes have been changed, but David changed the definition of `caml_sys_pagesize` to be 64Kib on Windows, which affects nothing else and lets people read the code as if we only ever asked for alignments dividing the page size and naturally round up the alignment value to that number.

gasche commented on runtime/unix.c:
```diff
@@ -478,3 +481,76 @@ void caml_init_os_params(void)
+#endif
+
+  return mem;
+}
```
Could you un-factor the Cygwin and the non-Cygwin versions as two separate definitions? It would be nice to explain in a comment why Cygwin does not need to care about alignment, possibly by referring to the `win32`-version comment.

dra27 replied:
Yes - will do!

gasche reviewed (commented):
The mmap stuff looks much nicer, thanks! I made a few extra suggestions there in comments.

dra27 commented:
Assertions added - I propose renaming `caml_sys_pagesize` to something which sounds less like it should be the pagesize.

gasche commented:
I have the feeling that I'm too picky -- eventually I'll send you commits myself -- but:

- Could the assertions not be debug assertions, but instead a normal conditional with a fatal error?
- I think it would make sense to track both the page size and the mmap granularity separately; with your change we round up the heap size to the mmap granularity, which has no particular performance justification other than "this used to be the page size". (Should they use `_plat_` or `_sys_`? If it was me they would both be `_sys_` and be defined in `sys.c` but I don't really care.)
- We could factor the unix.c code with a `caml_plat_mmap` function that does *not* take any alignment parameter. (This is exactly your Cygwin implementation, minus the alignment check.) Then the non-Cygwin path could call this function in the case were `alignment <= mmap_granularity`, and use the reserve-more-then-trim logic only in the second case. This makes, say, the GC or strace logs simpler in the minor-heap-reserve case. (It's not worth doing this if it costs code duplication, but it makes sense if we already implement the no-alignment version anyway.) If we do this right, the `reserve_only ? PROT_NONE : ...` logic can occur exactly once in the file instead of being duplicated as today.

dra27 commented:
> I have the feeling that I'm too picky -- eventually I'll send you commits myself -- but:
> 
> * Could the assertions not be debug assertions, but instead a normal conditional with a fatal error?

That's not you being picky, that was me not properly reading what you'd asked and then doing the opposite!! Changed to be `caml_fatal_error`!

> * I think it would make sense to track both the page size and the mmap granularity separately; with your change we round up the heap size to the mmap granularity, which has no particular performance justification other than "this used to be the page size". (Should they use `_plat_` or `_sys_`? If it was me they would both be `_sys_` and be defined in `sys.c` but I don't really care.)

I don't follow this - why would we track the page size when we never use it? There's no rounding up taking place - it's only the declaration of the alignment which is changing here? We're never using these functions to reserve anything close to as small as the page size or the granularity?

> * We could factor the unix.c code with a `caml_plat_mmap` function that does _not_ take any alignment parameter. (This is exactly your Cygwin implementation, minus the alignment check.) Then the non-Cygwin path could call this function in the case were `alignment <= mmap_granularity`, and use the reserve-more-then-trim logic only in the second case. This makes, say, the GC or strace logs simpler in the minor-heap-reserve case. (It's not worth doing this if it costs code duplication, but it makes sense if we already implement the no-alignment version anyway.) If we do this right, the `reserve_only ? PROT_NONE : ...` logic can occur exactly once in the file instead of being duplicated as today.

Possibly, although I was struggling to find the documented guarantee that `mmap` _always_ aligns to the page size for `MAP_FIXED` when passed addr of `0`? My reading is that if you _pass_ an explicit `addr` then it must be aligned to the page-size, but POSIX doesn't seem to require that the system obey that if _it_ allocates the address (it's quite explicit that the alignment requirement is for a non-zero addr)? May well just be my (mis)reading of it, though. We absolutely _know_ that `VirtualAlloc` will return a 64KiB-aligned address because it is defined in its documentation that it _will_ always do that, but I couldn't see an equivalently unambiguous declaration that all implementations of `mmap` will definitely return an address aligned to `_SC_PAGESIZE` in the same way?

gasche commented:
> I don't follow this - why would we track the page size when we never use it? There's no rounding up taking place - it's only the declaration of the alignment which is changing here? We're never using these functions to reserve anything close to as small as the page size or the granularity?

In

https://github.com/ocaml/ocaml/blob/1d8abf39c014a735494e3fbfb6750d898919b376/runtime/domain.c#L366-L373

we normalize the minor heap size to always contain a full number of memory pages, and I believe that the intention is really to use pages here rather than a granularity. In a recent past, other versions of the code had similar logic for much smaller sizes:

https://github.com/ocaml/ocaml/blob/87cd240f87e0c0527da15230dc913c76f2a74623/runtime/domain.c#L359

>  I was struggling to find the documented guarantee that mmap always aligns to the page size for MAP_FIXED when passed addr of 0? My reading is that if you pass an explicit addr then it must be aligned to the page-size, but POSIX doesn't seem to require that the system obey that if it allocates the address (it's quite explicit that the alignment requirement is for a non-zero addr)?

I don't know where to find reliable, portable information on this. My linux manpages are very clear:

> If addr is NULL, then the kernel chooses the (page-aligned) address at which to create the mapping; this is the most portable method of creating a new mapping. If addr is not NULL, then the kernel takes it as  a  hint  about where  to  place the mapping; on Linux, the kernel will pick a nearby page boundary

It's of course fine to leave the code as-is (it is tested and works) if we have doubts about this.

dra27 committed:
Introduce caml_sys_mmap_alignment

On Unix platforms, equal caml_sys_pagesize but on Windows reflects the
fact that VirtualAlloc has a stricter alignment guarantee.

dra27 commented:
Ah, I do follow now, thanks! A couple of extra commits pushed - I agree with the `caml_sys_pagesize` having the wrong prefix (turns out that was me in https://github.com/ocaml-multicore/ocaml-multicore/pull/173/commits/ac6280f1b968ce6bf5113390fa61e46987ff9f72!). I think it's OK for those to be in `platform.c` (there's a whole question for the future about organisation between `sys.c`, `platform.c`, `win32.c` and `unix.c`, but that can probably be for another day).

I would prefer to leave the Unix mmap part as-is - I agree that Linux is clear, but we then have to extend that to macOS, the other BSDs, etc. I guess it could be worth generalising the Cygwin approach to "platforms which guarantee mmap alignment", but I'm not sure that that gain is worth the risk?

What do you think of the present state of the PR, and I may have a look at tidying the commit history?

gasche commented on runtime/caml/platform.h:
```diff
@@ -127,6 +127,11 @@ void caml_plat_cond_free(caml_plat_cond*);
+   caml_plat_pagesize. The size given to caml_mem_unmap and caml_mem_decommit
+   must match the size given to caml_mem_map/caml_mem_commit for mem. The
+   alignment given to caml_mem_map must not be greater than
+   caml_plat_mmap_alignment on Windows and Cygwin. */
```
Suggestion:

```
/* The size given to caml_mem_map and caml_mem_commit must be a multiple of
   caml_plat_pagesize. The size given to caml_mem_unmap and caml_mem_decommit
   must match the size given to caml_mem_map/caml_mem_commit for mem.

   Our Windows and Cygwin implementations do not support arbitrary alignment,
   they will fail for [alignment] values greater than [caml_plat_mmap_alignment].
   Luckily, this value is rather large on those platforms: 64KiB. This is enough
   for all alignments used in the runtime system so far, the larger being our
   major heap pools aligned on 32KiB boundaries. */
```

gasche commented on runtime/caml/runtime_events.h:
```diff
@@ -226,7 +226,7 @@ CAMLextern char_os* caml_runtime_events_current_location();
 void caml_ev_end(ev_runtime_phase phase);
 void caml_ev_counter(ev_runtime_counter counter, uint64_t val);
-void caml_ev_lifecycle(ev_lifecycle lifecycle, int64_t data);
+CAMLextern void caml_ev_lifecycle(ev_lifecycle lifecycle, int64_t data);
```
it's a bit weird, why is only one function CAMLextern and not the others?

dra27 replied:
To answer it in two parts: it is marked `CAMLextern` because it is used (via the `CAML_EV_LIFECYCLE` macro) in `otherlibs/unix/fork.c` and has to be marked `__declspec(dllimport)` in order for FlexDLL to be able to relocate `dllunix.so`. It affects Cygwin only (and not also mingw-w64 in 5.0) just because mingw-w64 implementation doesn't compile `fork.c`.

I hadn't marked the others because it's all tied into the work I previously did trying to get towards using fvisiblity and that - it turns out - got tied into an obscure little bit of the relocatable work (full explanation is a bit big for the margins of this GitHub comment...). So I'd left it as being the minimal fix needed to get Cygwin64 back up for now! I don't mind speculatively going through the headers and adding `CAMLextern`; it's just not strictly related to the PR subject of restoring the Cygwin port.

gasche replied:
I would prefer if either all those `caml_ev_` functions had the same status (so yes, adding CAMLextern to all of them would be fine), or if there was a comment in the code to explain the inconsistency (and I think this is not as good as being consistent in the first place if there is no obvious downside to consistency).

dra27 replied:
I've added a comment - adding `__declspec(dllimport)` is not a zero-cost operation, although the performance impact depends on how hot the function calls are (cf. the Windows TLS PRs). I have previously agreed that `CAMLextern` is an unfortunate macro name (especially as function declarations don't require the `extern` keyword anyway) - `CAML_PUBLIC_API` or `CAML_NEEDS_TO_BE_VISIBLE_FROM_SHARED_OBJECTS` removes the link from C's `extern` keyword!

gasche replied:
In the present case, it is unclear to me why we would want to make just one function visible here and not the other (I would say they are all equally useful to users of our public api), so even if there is a cost the argument could be made that they should all be exported, or someone will be limited later and have to complain to ask us to export more. But if you think that minimal export is the right default, and there is a comment to explain that choice, then it is fine with me.

dra27 replied:
This isn't public API, though - these are all guarded by `CAML_INTERNALS` (sorry, the alternate suggestion `CAML_PUBLIC_API` probably masked that fact!)

gasche commented:
> I would prefer to leave the Unix mmap part as-is

Completely fine with me.

> What do you think of the present state of the PR, and I may have a look at tidying the commit history?

I started a reply on the fact that I have only looked at the memory-reservation part of the PR and nothing else, so I cannot comment on the rest. In fact, there is almost no rest: the memory-reservation part is the majority of the change. I haven't reviewed it in full yet, but I can say that I like the overall shape of the code, it is easier to read in particular that the current implementation. So tidying sounds right to me.

dra27 commented:
Branch tidied up, and it's running (again) through [precheck#796](https://ci.inria.fr/ocaml/job/precheck/796/).

The memory model tests hang on precheck, which is why I've left https://github.com/ocaml/ocaml/pull/11642/commits/8607dee068155bc0a1d62e135102e5581ca7e6b9 in, but it's a naff way of doing this (especially as it at present disables 3 additional tests). I'm just testing something else before replacing that commit with something less ad hoc.

xavierleroy commented on ocamltest/builtin_actions.ml:
```diff
@@ -110,7 +110,7 @@ let not_windows = make
 let is_bsd_system s =
   match s with
-  | "bsd_elf" | "netbsd" | "freebsd" | "openbsd" -> true
+  | "bsd_elf" | "netbsd" | "freebsd" | "openbsd" | "cygwin" -> true
```
That's a bold claim.  In tests where it matters, could we test "is_bsd_system or is_cygwin" instead?

xavierleroy replied:
Oh, OK, I see this is a temporary commit.

xavierleroy commented:
Side remark:
> I changed caml_sys_pagesize to be intnat rather than long in preference to having to add ARCH_LONG_PRINTF_FORMAT definitions.

The format for `long` is always `l`.  However, on Win64, `long` is still 32 bits, so either you can use `int` (because the quantity fits in 32 bits) or you should use `intnat` (to get a full machine word).

xavierleroy commented:
> I was struggling to find the documented guarantee that mmap always aligns to the page size for MAP_FIXED when passed addr of 0? My reading is that if you pass an explicit addr then it must be aligned to the page-size, but POSIX doesn't seem to require that the system obey that if it allocates the address (it's quite explicit that the alignment requirement is for a non-zero addr)?

I looked at some standard pages and man pages, and, yes, they are not very explicit about this.  One case where  mmap may return a non-page-aligned address is if it is given an offset into a file that is not page-aligned.  Under Linux this is an error, but OpenBSD seems to support it and says:
```
     If an fd and offset are specified, the resulting address may end up not
     on a page boundary, in order to align the page offset in the addr to the
     page offset in offset.
```

I think it's a reasonable assumption to make that `mmap(NULL, len, prot, MMAP_ANONYMOUS | flags, -1, 0)` always returns a page-aligned address.  However, the current code doesn't need this assumption, right?

xavierleroy commented on runtime/unix.c:
```diff
@@ -475,6 +478,108 @@ int caml_num_rows_fd(int fd)
+  base = (uintnat)mem;
+  aligned_start = (base + alignment - 1) & ~(alignment - 1);
+  aligned_end = aligned_start + size;
+  caml_gc_message(0x1000, "munmap %" ARCH_INTNAT_PRINTF_FORMAT "d"
+                          " bytes at %p for heaps\n",
+                          aligned_start - base, (void*)base);
+  munmap(mem, aligned_start - base);
+  caml_gc_message(0x1000, "munmap %" ARCH_INTNAT_PRINTF_FORMAT "d"
+                          " bytes at %p for heaps\n",
+                          (base + alloc_sz) - aligned_end, (void*)aligned_end);
+  munmap((void*)aligned_end, (base + alloc_sz) - aligned_end);
```
You should avoid calling `munmap` with length 0, i.e. if the address returned by `mmap` is suitably aligned already.  Not only this saves on system calls, but some standards say that `munmap` should fail if the length parameter is 0: https://pubs.opengroup.org/onlinepubs/9699919799/functions/munmap.html

dra27 replied:
Thanks for the review! The code in 5.0 is definitely calling `munmap` with length `0`, so a version of this should be extracted to go on 5.0. Would you prefer that I do a minimal patch, guarding the existing calls to `munmap`, or extract the fully split implementation from this PR? (@Octachron deferred to you 🙂)

xavierleroy replied:
For 5.0, I think we can either do nothing (hoping that all reasonable implementations of `munmap` accept zero lengths) or do a quick fix with a guard
```
   munmap(addr, len)    --->  if (len != 0) munmap(addr, len);
```
A backport from this PR feels too risky at this late point in the 5.0 release cycle.

xavierleroy reviewed (commented):
Globally, this looks good to me, and clearer than the earlier, Cygwin-incompatible implementation.  One remark below about unnecessary and potentially dangerous calls to `munmap`.  The ocamltest and test suite changes cannot be committed as is.  Otherwise, this looks good to go.

dra27 commented:
Rebased in light of #11726 - I've had a suggestion from @shindere for dealing with the ocamltest/testsuite part which I'll sort out and push next.

gasche commented on runtime/platform.c:
```diff
@@ -137,34 +138,28 @@ static uintnat round_up(uintnat size, uintnat align) {
-  void* mem;
-  uintnat base, aligned_start, aligned_end;
+  CAMLassert(Is_power_of_2(alignment));
+  CAMLassert(Is_page_aligned(size));
```
Reading the code again, I don't understand the reason for this assert (new in the PR) that `size` is page-aligned. It does not fail today, but what benefits does this restriction bring?

dra27 replied:
We never want to be mapping partial pages (just for the waste) - I thought that had come up in one or other of the various PRs on this, but I can't find a reference to it. It formalises something which is already the case and I think it simplifies the reasoning slightly for the Windows/Cygwin mapping, in that we know that the size we request is definitely what gets allocated. Sorry, it's a bit hand-wavy, but I'm fairly sure that's in the first commit on its own for a reason! 🙂

gasche commented on runtime/platform.c:
```diff
@@ -137,34 +138,28 @@ static uintnat round_up(uintnat size, uintnat align) {
-  uintnat base, aligned_start, aligned_end;
+  CAMLassert(Is_power_of_2(alignment));
+  CAMLassert(Is_page_aligned(size));
+  alignment = round_up(alignment, caml_plat_mmap_alignment);
```
It is non-intuitive to me why we don't, instead, when `alignment < caml_plat_mmap_alignment`, set `alignment = 0`, which would give us non-trimming behavior when possible. I think that the reason is that we are not 100% sure that `mmap` returns `caml_plat_mmap_alignment`-aligned addresses? Wouldn't it be worth documenting this surprising non-guarantee somewhere so that people understand why we are doing things the complex way?

gasche reviewed (approved):
Looks reasonable overall, here is an Approved mark. (I think you want to tweak some CI-related things before merging anyway?).

On re-read, I still had some noob questions again on alignment assumptions in the code -- our previous discussions didn't leave enough noob-friendly enough comments in the code that I could avoid being surprised again by what's going on in the details. Do what you want with my questions -- possibly nothing.

dra27 committed:
Skip the memory model tests on Jenkins for Cygwin

dra27 commented:
`OCAMLTEST_SKIP_TESTS` should now be being used to avoid the memory model tests for Cygwin, instead of the hack which was there before. This is running through a hopefully final [precheck](https://ci.inria.fr/ocaml/job/precheck/806/)

dra27 commented:
2 down, 1 to go...

xavierleroy commented:
Did you do any performance measurement for this resurrected Cygwin64 port?  Based on the Jenkins CI times, it seems to be about twice slower than the Mingw64 port, on similar VMs.  More precisely here are some timings for the "main" CI job:
- Cygwin64 VM: 2 hours (with -j2)
- Mingw64 VM: 1 hour (with -j2)
- x86-64 Linux server: 23 minutes (sequential)
- Mac Mini M1: 15 minutes (sequential).

xavierleroy commented:
Of course, this can also be an artefact of the VMs used for testing, e.g. Windows Defender going crazy.  So your own measurements will be good to know.
